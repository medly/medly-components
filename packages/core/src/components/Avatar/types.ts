import { AvatarSizes } from '@medly-components/theme';
import { HTMLProps, Omit } from '@medly-components/utils';

export interface AvatarProps extends Omit<HTMLProps<HTMLDivElement>, 'size'> {
    /** Size of the avatar */
    size?: AvatarSizes;
    /** Text color */
    textColor?: string;
    /** Background color */
    bgColor?: string;
    /** Hover Text color */
    hoverTextColor?: string;
    /** Hover Background color */
    hoverBgColor?: string;
    /** Hover text shadow color*/
    hoverTextShadowColor?: string;
    /** Hover image shadow color */
    hoverImgShadowColor?: string;
    /** Pressed Background color */
    pressedBgColor?: string;
    /** Pressed Text color */
    pressedTextColor?: string;
}

export interface StyledProps extends AvatarProps {
    /** To check children is image */
    isImage?: boolean;
}
