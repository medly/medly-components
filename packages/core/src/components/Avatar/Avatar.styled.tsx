import { css, styled, WithThemeProp } from '@medly-components/utils';
import Text from '../Text';
import { StyledProps } from './types';

const getAvatarSize = ({ theme: { avatar }, size }: StyledProps & WithThemeProp) => {
    const { sizes, defaults } = avatar;
    return sizes[size || defaults.size].avatarSize;
};

const getShadowColor = ({ theme: { avatar }, hoverTextShadowColor, hoverImgShadowColor, isImage }: StyledProps & WithThemeProp) => {
    return isImage
        ? hoverImgShadowColor || avatar.defaults.hover.imgShadowColor
        : hoverTextShadowColor || avatar.defaults.hover.textShadowColor;
};

const hoverStyle = ({ theme, hoverBgColor, hoverTextColor }: StyledProps & WithThemeProp) => {
    const { defaults } = theme.avatar;
    return css`
        &:hover {
            background: ${hoverBgColor || defaults.hover.bgColor};
            box-shadow: 0 0.4rem 0.8rem ${getShadowColor};
            ${Text.Style} {
                color: ${hoverTextColor || defaults.hover.textColor};
            }
        }
    `;
};

const pressedStyle = ({ theme, pressedBgColor, pressedTextColor }: StyledProps & WithThemeProp) => {
    const { defaults } = theme.avatar;
    return css`
        &:active {
            background: ${pressedBgColor || defaults.pressed.bgColor};
            ${Text.Style} {
                color: ${pressedTextColor || defaults.pressed.textColor};
            }
        }
    `;
};

const getTextStyle = ({ theme, ...props }: StyledProps & WithThemeProp) => {
    const { size, textColor, bgColor } = props,
        { defaults, sizes } = theme.avatar,
        { avatarSize, fontSize } = sizes[size || defaults.size];

    return css`
        color: ${textColor || defaults.textColor};
        background: ${bgColor || defaults.bgColor};
        user-select: none;

        ${Text.Style} {
            line-height: ${avatarSize};
            font-size: ${fontSize};
            font-weight: ${theme.font.weights[defaults.fontWeight]};
            font-family: ${defaults.fontFamily};
        }
    `;
};

export const AvatarStyled = styled('div').attrs(({ theme: { avatar: { defaults } } }: StyledProps & WithThemeProp) => ({
    defaults
}))<StyledProps>`
    display: inline-block;
    text-align: center;
    min-width: max-content;
    width: ${getAvatarSize};
    height: ${getAvatarSize};
    border-radius: ${({ defaults }) => defaults.borderRadius};
    overflow: hidden;
    cursor: ${({ onClick, disabled }) => (disabled ? 'not-allowed' : onClick ? 'pointer' : 'inherit')};

    ${getTextStyle}

    img {
        width: ${getAvatarSize};
        height: ${getAvatarSize};
        object-fit: cover;
        border: 0.1rem solid ${({ defaults }) => defaults.borderColor};
        box-sizing: border-box;
        border-radius: ${({ defaults }) => defaults.borderRadius};
    }

    ${props => props.onClick && hoverStyle};
    ${props => props.onClick && pressedStyle};
`;
