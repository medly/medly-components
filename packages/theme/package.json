{
    "name": "@medly-components/theme",
    "version": "6.0.0",
    "sideEffects": false,
    "main": "dist/cjs/index.js",
    "module": "dist/es/index.js",
    "types": "dist/es/index.d.ts",
    "files": [
        "dist"
    ],
    "publishConfig": {
        "access": "public"
    },
    "license": "MIT",
    "scripts": {
        "build": "cross-env BABEL_ENV=production rollup -c=../../rollup.config.js",
        "lint:fix": "yarn lint:ts --fix",
        "lint": "eslint 'src/**/*.{ts,tsx}'",
        "prebuild": "rimraf dist && tsc",
        "watch": "yarn run build -w"
    }
}
