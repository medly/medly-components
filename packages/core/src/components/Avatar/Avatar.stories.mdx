import { Avatar } from './Avatar.tsx';
import Text from '../Text';
import { defaultTheme } from '@medly-components/theme';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { sizes, Div, ThemeInterface } from './Avatar.stories';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta title="Core" component={Avatar} parameters={{ jest: ['Avatar.test.tsx'] }} />

# Avatar

The `Avatar` component displays profile pictures, images, icons or initials, and text representing people or other entities.

<Preview withToolbar>
    <Story name="Avatar" decorators={[withKnobs]}>
        <Div>
            <Avatar
                size={select('Size', sizes, 'M')}
                bgColor={color('Background color', defaultTheme.avatar.defaults.bgColor)}
                textColor={color('Text color', defaultTheme.avatar.defaults.textColor)}
                withHoverEffect={boolean('With Hover Effect', true)}
                hoverBgColor={color('Hover background color', defaultTheme.avatar.defaults.hover.bgColor)}
                hoverTextColor={color('Hover text color', defaultTheme.avatar.defaults.hover.textColor)}
                hoverImgShadowColor={color('Hover img shadow color', defaultTheme.avatar.defaults.hover.imgShadowColor)}
                hoverTextShadowColor={color('Hover text shadow color', defaultTheme.avatar.defaults.hover.textShadowColor)}
            >
                <img src="http://i.pravatar.cc/150?img=66" />
            </Avatar>
            <Avatar
                size={select('Size', sizes, 'M')}
                bgColor={color('Background color', defaultTheme.avatar.defaults.bgColor)}
                textColor={color('Text color', defaultTheme.avatar.defaults.textColor)}
                withHoverEffect={boolean('With Hover Effect', true)}
                hoverBgColor={color('Hover background color', defaultTheme.avatar.defaults.hover.bgColor)}
                hoverTextColor={color('Hover text color', defaultTheme.avatar.defaults.hover.textColor)}
                hoverImgShadowColor={color('Hover img shadow color', defaultTheme.avatar.defaults.hover.imgShadowColor)}
                hoverTextShadowColor={color('Hover text shadow color', defaultTheme.avatar.defaults.hover.textShadowColor)}
            >
                <Text>KP</Text>
            </Avatar>
        </Div>
    </Story>
</Preview>

## Color variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <Div>
        <Avatar
            sizes="L"
            textColor={defaultTheme.colors.green[500]}
            bgColor={defaultTheme.colors.green[100]}
            hoverBgColor={defaultTheme.colors.green[400]}
            hoverTextColor={defaultTheme.colors.white}
            hoverImgShadowColor="rgba(96, 120, 144, 0.35)"
            hoverTextShadowColor="rgba(0, 128, 0, 0.35)"
        >
            <Text>KP</Text>
        </Avatar>
        <Avatar
            sizes="L"
            textColor={defaultTheme.colors.blush[500]}
            bgColor={defaultTheme.colors.blush[100]}
            hoverBgColor={defaultTheme.colors.blush[400]}
            hoverTextColor={defaultTheme.colors.white}
            hoverImgShadowColor="rgba(96, 120, 144, 0.35)"
            hoverTextShadowColor="rgba(242, 113, 33, 0.35)"
        >
            <Text>KP</Text>
        </Avatar>
    </Div>
</Preview>

## Sizes

Three sizes are available: `S` | `M` | `L`.

<Preview withToolbar>
    <Div>
        <Avatar size="S">SM</Avatar>
        <Avatar size="M">MD</Avatar>
        <Avatar size="L">LG</Avatar>
        <Avatar size="S">
            <img src="http://i.pravatar.cc/150?img=66" />
        </Avatar>
        <Avatar size="M">
            <img src="http://i.pravatar.cc/150?img=66" />
        </Avatar>
        <Avatar size="L">
            <img src="http://i.pravatar.cc/150?img=66" />
        </Avatar>
    </Div>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={Avatar} />

## AvatarTheme

This component offers the following themes mentioned in the table below:

<Props of={ThemeInterface} />
