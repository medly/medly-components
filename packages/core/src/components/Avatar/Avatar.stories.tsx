import { AvatarTheme, defaultTheme } from '@medly-components/theme';
import { styled } from '@medly-components/utils';
import React from 'react';
import { AvatarProps } from './types';

export const sizes: Required<AvatarProps>['size'][] = ['S', 'M', 'L'];

export const ThemeInterface: React.FC<AvatarTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.avatar
};

export const Div = styled.div`
    & > * {
        margin-right: 10px;
    }
`;
