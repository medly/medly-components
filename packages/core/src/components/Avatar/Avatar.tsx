import { isValidStringOrNumber, WithStyle } from '@medly-components/utils';
import React, { FC } from 'react';
import Text from '../Text';
import { AvatarStyled } from './Avatar.styled';
import { AvatarProps } from './types';

const Component: FC<AvatarProps> = React.memo(
    React.forwardRef((props, ref) => {
        const isImage = !!React.Children.toArray(props.children).find(
            (child: any) => child.type === 'img' || child.props?.originalType === 'img'
        );
        return (
            <AvatarStyled ref={ref} {...props} isImage={isImage}>
                {React.Children.map(props.children, c => {
                    return isValidStringOrNumber(c) ? (
                        <Text uppercase textWeight="Strong">
                            {c}
                        </Text>
                    ) : (
                        c
                    );
                })}
            </AvatarStyled>
        );
    })
);
Component.displayName = 'Avatar';
export const Avatar: FC<AvatarProps> & WithStyle = Object.assign(Component, {
    Style: AvatarStyled
});
