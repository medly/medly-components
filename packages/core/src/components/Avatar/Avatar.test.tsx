import { defaultTheme } from '@medly-components/theme';
import { render } from '@test-utils';
import React from 'react';
import { Avatar } from './Avatar';

describe('Avatar component', () => {
    it('should render with default theme', () => {
        const { container } = render(<Avatar>M</Avatar>);
        expect(container).toMatchSnapshot();
    });

    it('should render with all the props given', () => {
        const { container } = render(
            <Avatar
                size="M"
                textColor={defaultTheme.colors.green[500]}
                bgColor={defaultTheme.colors.green[100]}
                hoverBgColor={defaultTheme.colors.green[400]}
                hoverTextColor={defaultTheme.colors.white}
                hoverImgShadowColor="rgba(96, 120, 144, 0.35)"
                hoverTextShadowColor="rgba(0, 128, 0, 0.35)"
                pressedBgColor="rgba(96, 120, 144, 0.65)"
                pressedTextColor="rgba(0, 128, 0, 0.65)"
                onClick={jest.fn()}
            >
                M
            </Avatar>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with default hover and pressed states', () => {
        const { container } = render(
            <Avatar size="M" textColor={defaultTheme.colors.green[500]} bgColor={defaultTheme.colors.green[100]} onClick={jest.fn()}>
                M
            </Avatar>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with not-allowed cursor when it is disabled', () => {
        const { container } = render(
            <Avatar size="M" disabled>
                M
            </Avatar>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render image avatar properly', () => {
        const { container } = render(
            <Avatar size="L" onClick={jest.fn()}>
                <img src="https://dummurl" />
            </Avatar>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render avatar without hover and pressed states', () => {
        const { container } = render(
            <Avatar
                size="M"
                textColor={defaultTheme.colors.green[500]}
                bgColor={defaultTheme.colors.green[100]}
                hoverBgColor={defaultTheme.colors.green[400]}
                hoverTextColor={defaultTheme.colors.white}
                hoverImgShadowColor="rgba(96, 120, 144, 0.35)"
                hoverTextShadowColor="rgba(0, 128, 0, 0.35)"
                pressedBgColor="rgba(96, 120, 144, 0.65)"
                pressedTextColor="rgba(0, 128, 0, 0.65)"
            >
                M
            </Avatar>
        );
        expect(container).toMatchSnapshot();
    });
});
