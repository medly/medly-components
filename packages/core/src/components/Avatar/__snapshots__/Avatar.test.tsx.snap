// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component should render avatar without hover 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 700;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  text-transform: uppercase;
}

.c0 {
  display: inline-block;
  text-align: center;
  width: 4rem;
  height: 4rem;
  border-radius: 50%;
  overflow: hidden;
  cursor: pointer;
  color: #008000;
  background: #E6F2E6;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 .c1 {
  line-height: 4rem;
  font-size: 1.6rem;
  font-weight: 400;
  font-family: Open Sans;
}

.c0 img {
  width: 4rem;
  height: 4rem;
  object-fit: cover;
  border: 0.1rem solid #dfe4e9;
  box-sizing: border-box;
  border-radius: 50%;
}

<div>
  <div
    class="c0"
  >
    <strong
      class="c1 c2"
    >
      M
    </strong>
  </div>
</div>
`;

exports[`Avatar component should render image avatar properly 1`] = `
.c0 {
  display: inline-block;
  text-align: center;
  width: 4.8rem;
  height: 4.8rem;
  border-radius: 50%;
  overflow: hidden;
  cursor: inherit;
  color: #126AFA;
  background: #E7F0FF;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 .sc-AxjAm {
  line-height: 4.8rem;
  font-size: 2rem;
  font-weight: 400;
  font-family: Open Sans;
}

.c0 img {
  width: 4.8rem;
  height: 4.8rem;
  object-fit: cover;
  border: 0.1rem solid #dfe4e9;
  box-sizing: border-box;
  border-radius: 50%;
}

.c0:hover {
  background: #70ADE9;
  box-shadow: 0 0.4rem 0.8rem rgba(96,120,144,0.35);
}

.c0:hover .sc-AxjAm {
  color: #ffffff;
}

<div>
  <div
    class="c0"
  >
    <img
      src="http://dummurl"
    />
  </div>
</div>
`;

exports[`Avatar component should render with all the props given 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 700;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  text-transform: uppercase;
}

.c0 {
  display: inline-block;
  text-align: center;
  width: 4rem;
  height: 4rem;
  border-radius: 50%;
  overflow: hidden;
  cursor: pointer;
  color: #008000;
  background: #E6F2E6;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 .c1 {
  line-height: 4rem;
  font-size: 1.6rem;
  font-weight: 400;
  font-family: Open Sans;
}

.c0 img {
  width: 4rem;
  height: 4rem;
  object-fit: cover;
  border: 0.1rem solid #dfe4e9;
  box-sizing: border-box;
  border-radius: 50%;
}

.c0:hover {
  background: #269326;
  box-shadow: 0 0.4rem 0.8rem rgba(0,128,0,0.35);
}

.c0:hover .c1 {
  color: #ffffff;
}

<div>
  <div
    class="c0"
  >
    <strong
      class="c1 c2"
    >
      M
    </strong>
  </div>
</div>
`;

exports[`Avatar component should render with default theme 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 700;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  text-transform: uppercase;
}

.c0 {
  display: inline-block;
  text-align: center;
  width: 4rem;
  height: 4rem;
  border-radius: 50%;
  overflow: hidden;
  cursor: inherit;
  color: #126AFA;
  background: #E7F0FF;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 .c1 {
  line-height: 4rem;
  font-size: 1.6rem;
  font-weight: 400;
  font-family: Open Sans;
}

.c0 img {
  width: 4rem;
  height: 4rem;
  object-fit: cover;
  border: 0.1rem solid #dfe4e9;
  box-sizing: border-box;
  border-radius: 50%;
}

.c0:hover {
  background: #70ADE9;
  box-shadow: 0 0.4rem 0.8rem rgba(0,90,238,0.35);
}

.c0:hover .c1 {
  color: #ffffff;
}

<div>
  <div
    class="c0"
  >
    <strong
      class="c1 c2"
    >
      M
    </strong>
  </div>
</div>
`;

exports[`Avatar component should render with not-allowed cursor when it is disabled 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 700;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  text-transform: uppercase;
}

.c0 {
  display: inline-block;
  text-align: center;
  width: 4rem;
  height: 4rem;
  border-radius: 50%;
  overflow: hidden;
  cursor: not-allowed;
  color: #126AFA;
  background: #E7F0FF;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 .c1 {
  line-height: 4rem;
  font-size: 1.6rem;
  font-weight: 400;
  font-family: Open Sans;
}

.c0 img {
  width: 4rem;
  height: 4rem;
  object-fit: cover;
  border: 0.1rem solid #dfe4e9;
  box-sizing: border-box;
  border-radius: 50%;
}

<div>
  <div
    class="c0"
    disabled=""
  >
    <strong
      class="c1 c2"
    >
      M
    </strong>
  </div>
</div>
`;
